File: src/Fuaran.UI.VisualBasic.Conformance.Tests/Conformance.vb
Imports System.IO
Imports System.Text.Json
Imports Fuaran.UI.VisualBasic
Imports Csharp = Fuaran.UI.CSharp

' Phase 312 — corpus conformance by construction (the VB analogue of Phase 306).
' Drives the full shared wire-format-fixtures corpus through the veneer:
'
'   * every node / op round-trip fixture decodes + re-encodes byte-identically
'     through the C#-native Decode facade the VB veneer reuses (no VB codec);
'   * every reject fixture surfaces the canonical code + path;
'   * coverage-vs-corpus: every node fixture's kind.$type has a VB XML element.
'
' ARIA reconciliation (Phase 313, same decision as C# Phase 307): the corpus is bare;
' the factories inherit ARIA, so authoring a fixture and byte-comparing to the bare
' file only lands for the ARIA-free kinds (see Authoring.vb). The decode round-trip
' proves byte-fidelity ARIA-agnostically. No corpus regen, no fourth CI leg.
Module Conformance

    Sub Run(h As Harness)
        If Not Corpus.Available Then
            Console.WriteLine("[conformance] wire-format-fixtures corpus absent — skipping full corpus run.")
            Return
        End If

        Using doc = JsonDocument.Parse(File.ReadAllText(Path.Combine(Corpus.Root, "manifest.json")))
            For Each f In doc.RootElement.GetProperty("fixtures").EnumerateArray()
                Dim id = f.GetProperty("id").GetString()
                Dim kind = f.GetProperty("kind").GetString()
                Dim decoder = If(GetStr(f, "decoder"), "node")
                Dim input = Corpus.ReadFixture(f.GetProperty("inputFile").GetString())

                Select Case kind
                    Case "node-round-trip"
                        h.Check($"node round-trip {id}", Csharp.Decode.NodeRoundTrips(input),
                                "decode/re-encode not byte-identical")
                        Dim kt = KindType(input)
                        h.Check($"coverage-vs-corpus: <{Element(kt)}> ({id})",
                                kt IsNot Nothing AndAlso FuaranXml.IsKnownElement(Element(kt)),
                                $"no VB element authors kind {kt}")
                    Case "op-round-trip"
                        h.Check($"op round-trip {id}", Csharp.Decode.OpRoundTrips(input),
                                "decode/re-encode not byte-identical")
                    Case "reject"
                        RunReject(h, id, decoder, input, GetStr(f, "expectedErrorCode"), GetStr(f, "expectedPath"))
                End Select
            Next
        End Using
    End Sub

    Private Sub RunReject(h As Harness, id As String, decoder As String, input As String,
                          expectedCode As String, expectedPath As String)
        Dim err = If(decoder = "op", Csharp.Decode.Op(input).Error, Csharp.Decode.Node(input).Error)
        If err Is Nothing Then
            h.Check($"reject {id}", False, "decoded ok — expected a reject")
            Return
        End If
        h.Check($"reject {id}: code {expectedCode}", err.Code = expectedCode, $"got {err.Code}")
        h.Check($"reject {id}: path {expectedPath}",
                expectedPath Is Nothing OrElse err.Path.StartsWith(expectedPath), $"got {err.Path}")
    End Sub

    ' LayoutKind.GridLayout is authored as <Grid>; every other kind name = element name.
    Private Function Element(kindType As String) As String
        Return If(kindType = "GridLayout", "Grid", kindType)
    End Function

    ' Parse corpus payloads at the wire format's own syntactic bound, not
    ' System.Text.Json's 64-level default. A node level costs several JSON
    ' levels, so the §21 accept fixture at exactly MaxDepth node levels is 72
    ' JSON levels — well inside what the decoder accepts and past what the
    ' default reader will read. On the default this harness did not fail a
    ' fixture, it THREW, killing the whole conformance run at the first deep
    ' payload.
    Private ReadOnly WireJson As New JsonDocumentOptions With {
        .MaxDepth = Global.Fuaran.UI.WireLimits.MaxJsonDepth
    }

    Private Function KindType(nodeJson As String) As String
        Using doc = JsonDocument.Parse(nodeJson, WireJson)
            Dim kind As JsonElement
            If Not doc.RootElement.TryGetProperty("kind", kind) Then Return Nothing
            Dim t As JsonElement
            If Not kind.TryGetProperty("$type", t) Then Return Nothing
            Return t.GetString()
        End Using
    End Function

    Private Function GetStr(el As JsonElement, name As String) As String
        Dim v As JsonElement
        If el.TryGetProperty(name, v) Then Return v.GetString()
        Return Nothing
    End Function

End Module

File: src/Fuaran.UI.CSharp.Conformance.Tests/Conformance.cs
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace Fuaran.UI.CSharp.Conformance.Tests;

// Phase 306 — corpus conformance by construction. Drives the full shared
// wire-format-fixtures corpus through the veneer:
//
//   * every node / op round-trip fixture decodes + re-encodes byte-identically
//     through the C#-native Decode facade (Phase 309), proving the veneer lands
//     on the canonical wire — no fourth CI leg, the byte-canonicity rides the
//     existing F#<->corpus gate;
//   * every reject fixture surfaces the canonical code + path;
//   * coverage-vs-corpus: every node fixture's kind.$type has a C# authoring
//     factory, so a corpus that grows a kind the veneer can't author fails here.
//
// The ARIA reconciliation (Phase 307): the corpus is authored BARE (no ARIA),
// whereas the C# factories inherit per-component ARIA from the F# smart ctors, so
// authoring a fixture via the factories and byte-comparing to the bare file would
// diverge for ARIA-bearing kinds. The decode round-trip proves byte-fidelity
// ARIA-agnostically (bare in -> bare out); the C#==F#-smart-ctor authoring parity
// is proven separately (Program.cs / Accessibility.cs). No corpus regen.
internal static class Conformance
{
    private sealed record Fixture(string Id, string Kind, string Decoder, string InputFile, string? ExpectedFile, string? ExpectedErrorCode, string? ExpectedPath);

    public static void Run(Harness h)
    {
        if (!Corpus.Available)
        {
            System.Console.WriteLine("[conformance] wire-format-fixtures corpus absent — skipping full corpus run.");
            return;
        }

        foreach (var f in LoadManifest())
        {
            switch (f.Kind)
            {
                case "node-round-trip":
                    RunNodeRoundTrip(h, f);
                    break;
                case "op-round-trip":
                    RunOpRoundTrip(h, f);
                    break;
                case "reject":
                    RunReject(h, f);
                    break;
            }
        }
    }

    private static void RunNodeRoundTrip(Harness h, Fixture f)
    {
        var input = Corpus.ReadFixture(f.InputFile);
        var expected = Corpus.ReadFixture(f.ExpectedFile ?? f.InputFile);

        if (!Decode.TryNode(input, out var node, out var err))
        {
            h.Check($"node round-trip {f.Id}", false, $"decode failed: {err.Code} @ {err.Path}");
            return;
        }

        h.ByteEqual($"node round-trip {f.Id}", node.Encode(), expected);

        // Coverage-vs-corpus: the fixture's kind.$type must have a C# authoring factory.
        var kindType = KindType(input);
        h.Check(
            $"coverage-vs-corpus: factory for {kindType} ({f.Id})",
            kindType is not null && Coverage.HasFactoryForKind(kindType),
            kindType is null ? "(no kind.$type in fixture)" : $"no C# factory authors kind {kindType}");
    }

    private static void RunOpRoundTrip(Harness h, Fixture f)
    {
        var input = Corpus.ReadFixture(f.InputFile);
        var expected = Corpus.ReadFixture(f.ExpectedFile ?? f.InputFile);

        if (!Decode.TryOp(input, out var op, out var err))
        {
            h.Check($"op round-trip {f.Id}", false, $"decode failed: {err.Code} @ {err.Path}");
            return;
        }

        h.ByteEqual($"op round-trip {f.Id}", op.Encode(), expected);
    }

    private static void RunReject(Harness h, Fixture f)
    {
        var input = Corpus.ReadFixture(f.InputFile);
        var error = f.Decoder == "op" ? Decode.Op(input).Error : Decode.Node(input).Error;

        if (error is null)
        {
            h.Check($"reject {f.Id}", false, "decoded ok — expected a reject");
            return;
        }

        h.Check(
            $"reject {f.Id}: code {f.ExpectedErrorCode}",
            error.Code == f.ExpectedErrorCode,
            $"expected {f.ExpectedErrorCode}, got {error.Code}");

        h.Check(
            $"reject {f.Id}: path {f.ExpectedPath}",
            f.ExpectedPath is null || error.Path.StartsWith(f.ExpectedPath),
            $"expected path prefix {f.ExpectedPath}, got {error.Path}");
    }

    private static IEnumerable<Fixture> LoadManifest()
    {
        using var doc = JsonDocument.Parse(File.ReadAllText(Path.Combine(Corpus.Root!, "manifest.json")));
        foreach (var el in doc.RootElement.GetProperty("fixtures").EnumerateArray())
        {
            yield return new Fixture(
                el.GetProperty("id").GetString()!,
                el.GetProperty("kind").GetString()!,
                el.TryGetProperty("decoder", out var d) ? d.GetString()! : "node",
                el.GetProperty("inputFile").GetString()!,
                el.TryGetProperty("expectedFile", out var e) ? e.GetString() : null,
                el.TryGetProperty("expectedErrorCode", out var c) ? c.GetString() : null,
                el.TryGetProperty("expectedPath", out var p) ? p.GetString() : null);
        }
    }

    // Parse corpus payloads at the wire format's own syntactic bound, not
    // System.Text.Json's 64-level default. A node level costs several JSON
    // levels, so the §21 accept fixture at exactly `MaxDepth` node levels is 72
    // JSON levels — well inside what the decoder accepts and past what the
    // default reader will read. On the default this harness did not fail a
    // fixture, it THREW, killing the whole conformance run at the first deep
    // payload.
    private static readonly JsonDocumentOptions WireJson = new()
    {
        MaxDepth = global::Fuaran.UI.WireLimits.MaxJsonDepth,
    };

    private static string? KindType(string nodeJson)
    {
        using var doc = JsonDocument.Parse(nodeJson, WireJson);
        return doc.RootElement.TryGetProperty("kind", out var kind) && kind.TryGetProperty("$type", out var t)
            ? t.GetString()
            : null;
    }
}
